Implement multiple import maps

yoavweiss opened the pull request:
#### c363df60bf5834d874f9830efeebc3892b87ff25
<pre>
Implement multiple import maps

<a href="https://bugs.webkit.org/show_bug.cgi?id=279025">https://bugs.webkit.org/show_bug.cgi?id=279025</a>

Reviewed by Yusuke Suzuki.

This PR enables multiple import maps to co-exist in a single document,
and to be loaded after a module script was loaded.

It also removes some code related to external import maps, which was
removed at <a href="https://github.com/WebKit/WebKit/pull/37629.">https://github.com/WebKit/WebKit/pull/37629.</a>

* LayoutTests/TestExpectations: Remove skip and send console to stderr.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/dynamic-import-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/script-tag-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/script-tag-inline-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare-specifiers.sub-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target-dynamic.sub-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target.sub-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/hash-failure-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/nonce-failure-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_data-url-prefix.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_empty-import-map.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_overlapping-entries.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_packages-via-trailing-slashes.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_resolving-null.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_scopes-exact-vs-prefix.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_tricky-specifiers.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_url-specifiers-schemes.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_url-specifiers.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-url-specifiers.sub-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/http-url-like-specifiers.sub-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/already-resolved-dropped-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/basic-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/conflict-first-persists-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/resolution-consistency-in-module-tree-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/resolution-consistency-in-module-tree-inline-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/resolution-consistency-in-module-tree.html: Avoid reliance on async inline module timing.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/url-resolution-conflict-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/with-errors-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-as-classic-script-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/dynamic-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/external-script-bare-descendent-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/external-script-descendent-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/external-script-descendent-with-scope-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/failed-resolution-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/integrity-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/prefix-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/script-descendent-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/url-resolution-conflict-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/resources/out-of-scope-test.js: Added. Missed during import.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/import-maps/bare-specifiers.sub-expected.txt: progression.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target-dynamic.sub-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target.sub-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/import-maps/data-url-specifiers.sub-expected.txt: progression.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/import-maps/http-url-like-specifiers.sub-expected.txt: progression.
* LayoutTests/platform/win/imported/w3c/web-platform-tests/import-maps/bare-specifiers.sub-expected.txt: progression.
* LayoutTests/platform/win/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target-dynamic.sub-expected.txt: Removed.
* LayoutTests/platform/win/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target.sub-expected.txt: Removed.
* LayoutTests/platform/win/imported/w3c/web-platform-tests/import-maps/data-url-specifiers.sub-expected.txt: progression.
* LayoutTests/platform/win/imported/w3c/web-platform-tests/import-maps/http-url-like-specifiers.sub-expected.txt: progression.
* Source/JavaScriptCore/builtins/BuiltinNames.h: remove importMapStatus.
* Source/JavaScriptCore/builtins/ModuleLoader.js: remove
  waiting on importMapStatus.
(visibility.PrivateRecursive.async loadModule):
(visibility.PrivateRecursive.async loadAndEvaluateModule):
(visibility.PrivateRecursive.async requestImportModule):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h: remove importMapStatus.
* Source/JavaScriptCore/runtime/ImportMap.cpp: Added multiple import map
  logic.
(JSC::ImportMapInternal::mergeModuleSpecifierMaps): <a href="https://html.spec.whatwg.org/C#merge-module-specifier-maps">https://html.spec.whatwg.org/C#merge-module-specifier-maps</a>
(JSC::ImportMap::ImportMap): Constructor to enable creation through
a static function.
(JSC::ImportMap::resolve): refactor and add a call to
addModuleToResolvedModuleSet.
(JSC::normalizeSpecifierKey): Change the reporter to a reference.
(JSC::sortAndNormalizeSpecifierMap): Change the reporter to a reference.
(JSC::ImportMap::parseImportMapString): a static function creating
a new ImportMap from an import map string. Replaces registerImportMap.
(JSC::ImportMap::mergeExistingAndNewImportMaps): <a href="https://html.spec.whatwg.org/C#merge-existing-and-new-import-maps">https://html.spec.whatwg.org/C#merge-existing-and-new-import-maps</a>
the scopes map. The scopes vector is then used to iterate over the
scopes in order.
(JSC::findUrlPrefixes): Find all the prefixes of a URL
(JSC::ImportMap::addModuleToResolvedModuleSet): <a href="https://html.spec.whatwg.org/C#add-module-to-resolved-module-set">https://html.spec.whatwg.org/C#add-module-to-resolved-module-set</a>
(JSC::ImportMap::resolve const): Deleted.
(JSC::ImportMap::registerImportMap): Deleted.
* Source/JavaScriptCore/runtime/ImportMap.h: Replace the previous Scopes
  data structure with a HashMap and a Vector, used for fast retrival and
  in-order iteration.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp: Remove isAcquiring
  and pendingImportMap logic.
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::isAcquiringImportMaps const): Deleted.
(JSC::JSGlobalObject::visitChildrenImpl):
(JSC::JSGlobalObject::setAcquiringImportMaps): Deleted.
(JSC::JSGlobalObject::setPendingImportMaps): Deleted.
(JSC::JSGlobalObject::clearPendingImportMaps): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::importMap):
(JSC::JSGlobalObject::importMapStatusPromise const): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp: Remove
  importMapStatus.
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.h:
* Source/WebCore/bindings/js/ScriptController.cpp: Remove isAcquiring
  and pendingImportMap logic. Add error reporting to
  ImportMapLogReporter. Call mergeExistingAndNewImportMaps.
(WebCore::ScriptController::isAcquiringImportMaps): Deleted.
(WebCore::ScriptController::registerImportMap):
(WebCore::ScriptController::setAcquiringImportMaps): Deleted.
(WebCore::ScriptController::setPendingImportMaps): Deleted.
(WebCore::ScriptController::clearPendingImportMaps): Deleted.
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp: Remove spurious
  include.
* Source/WebCore/dom/LoadableClassicScript.cpp:
* Source/WebCore/dom/ScriptElement.cpp: Remove logic firing an error
  event related to isAcquiring. Remove pendingImportMap logic.
(WebCore::ScriptElement::determineScriptType):
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::registerImportMap):
(WebCore::ScriptElement::executePendingScript):
(WebCore::ScriptElement::requestImportMap): Deleted.
* Source/WebCore/dom/ScriptType.h: Fix up spec comments.

Canonical link: <a href="https://commits.webkit.org/288358@main">https://commits.webkit.org/288358@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8abdfb0b428278b7d2bafc9d0110975a9d8f125

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82789 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2438 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36993 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87907 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33846 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84882 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2510 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10352 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64506 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22272 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85845 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1863 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75303 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44782 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1742 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29491 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32879 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75773 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73000 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30163 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89274 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81836 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10080 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7332 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72922 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10308 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71111 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72137 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16300 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1468 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12818 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10033 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15554 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/104241 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9907 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25263 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13372 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11676 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -34,6 +34,38 @@
+            message.append("An import map rule for specifier '"_s);
+            message.append(specifier);
+            message.append("' was removed, as it conflicted with an existing rule."_s);
+            reporter.reportWarning(message.toString());
```
Let's just use `reporter.reportWarning(makeString(("An import map rule for specifier '"_s, specifier, "' was removed, as it conflicted with an existing rule."_s));`

yoavweiss replied:
Done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -34,6 +34,38 @@
+    // Instead of copying the maps and returning the copy, we're modifying the
+    // maps in place.
+    // 2. For each specifier → url of newMap:
+    for (auto specifier : newMap.keys()) {
```
Let's take it via `auto&`

yoavweiss replied:
Done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -34,6 +34,38 @@
+        }
+        auto url = newMap.get(specifier);
+        // 2.2. Set mergedMap[specifier] to url.
+        oldMap.add(specifier, url);
```
Let's make it
```
oldMap.add(specifier, WTFMove(url));
```

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -91,48 +123,48 @@ static URL parseURLLikeModuleSpecifier(const String& specifier, const URL& baseU
+        if (scope == baseURL || (scope.string().endsWith('/') && baseURL.string().startsWith(scope.string()))) {
             dataLogLnIf(ImportMapInternal::verbose, "        Matching");
-            auto result = resolveImportMatch(normalizedSpecifier, asURL, entry.m_map);
+            auto map = m_scopesMap.get(scope);
```
This copies the `map` and inefficient. So, let's do,
```
auto result = resolveImportMatch(normalizedSpecifier, asURL, m_scopesMap.get(scope));
```

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -153,88 +185,90 @@ static ImportMap::SpecifierMap sortAndNormalizeSpecifierMap(Ref<JSON::Object> im
             }
 
-            scopes.append({ scopePrefixURL, sortAndNormalizeSpecifierMap(potentialSpecifierMap.releaseNonNull(), baseURL, reporter) });
+            scopesMap.set(scopePrefixURL, sortAndNormalizeSpecifierMap(potentialSpecifierMap.releaseNonNull(), baseURL, reporter));
```
Use `WTFMove(scopePrefixURL)`

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+    ImportMap::IntegrityMap& newImportMapIntegrity = newImportMap->m_integrity;
+
+    // 3. For each scopePrefix → scopeImports of newImportMapScopes:
+    for (auto scope : newImportMapScopes) {
```
Use `auto&`.

yoavweiss replied:
done!

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+        m_toplevelResolvedModuleSet.add(specifierPrefix);
+
+    Vector<AtomString> referringScriptPrefixes = findURLPrefixes(referringScriptURL);
+    for (AtomString referringScriptPrefix : referringScriptPrefixes) {
```
Use `auto&`.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+    AtomString atomSpecifier = AtomString(specifier);
+    m_toplevelResolvedModuleSet.add(atomSpecifier);
+    Vector<AtomString> specifierPrefixes = findURLPrefixes(specifier);
+    for (AtomString specifierPrefix : specifierPrefixes)
```
Use `auto&`.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+        // specifiers whose prefix is in the set.
+        const auto currentSetIt = m_scopedResolvedModuleMap.getOptional(AtomString(scope.key.string()));
+        if (currentSetIt) {
+            const auto currentResolvedSet = currentSetIt.value();
```
Use `find`.
```
auto iter = m_scopedResolvedModuleMap.find(AtomString(scope.key.string()));
if (iter != m_scopedResolvedModuleMap.end()) {
   auto& currentResolvedSet = iter->value;
}
```

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+        if (currentSetIt) {
+            const auto currentResolvedSet = currentSetIt.value();
+            Vector<String> specifiersToRemove;
+            for (auto specifier : scopeImports.keys()) {
```
Use `auto&`.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+                if (currentResolvedSet.find(AtomString(specifier)) != currentResolvedSet.end())
+                    specifiersToRemove.append(specifier);
+            }
+            for (auto specifier : specifiersToRemove) {
```
Use `auto&`.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+                message.append("An import map scope rule for specifier '"_s);
+                message.append(specifier);
+                message.append("' was removed, as it conflicted with already resolved module specifiers."_s);
+                reporter.reportWarning(message.toString());
```
Use `makeString`.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+            const auto currentResolvedSet = currentSetIt.value();
+            Vector<String> specifiersToRemove;
+            for (auto specifier : scopeImports.keys()) {
+                if (currentResolvedSet.find(AtomString(specifier)) != currentResolvedSet.end())
```
Using `AtomString(...)` again and again is inefficient. If we need to do this frequently, should we store AtomString from the beginning in the scopeImports etc.?

yoavweiss replied:
Done!

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
 }
 
+// https://html.spec.whatwg.org/C/#merge-existing-and-new-import-maps
+void ImportMap::mergeExistingAndNewImportMaps(ImportMap* newImportMap, const ImportMap::Reporter& reporter)
```
I wonder if we should just take this as `Ref<ImportMap>&& newImportMap` and destroy the content.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+            // 3.3 Otherwise, set oldImportMap's scopes[scopePrefix] to
+            // scopeImports.
+            m_scopesMap.set(scope.key, scopeImports);
+            m_scopesVector.append(scope.key);
```
If we take newImportMap as move semantics, we no longer care about the newImportMap's content. So we can use `WTFMove(scope.value)`?

yoavweiss replied:
`WTFMove`d `scopeImports`

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+    }
+
+    // 4. For each url → integrity of newImportMap's integrity:
+    for (auto url : newImportMapIntegrity.keys()) {
```
Ditto.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+
+    // 4. For each url → integrity of newImportMap's integrity:
+    for (auto url : newImportMapIntegrity.keys()) {
+        auto newIntegrityValue = newImportMapIntegrity.get(url);
```
Ditto. It is copying.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+            message.append("An import map integrity rule for url '"_s);
+            message.append(url.string());
+            message.append("' was removed, as it conflicted with already defined integrity rules."_s);
+            reporter.reportWarning(message.toString());
```
Ditto.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+    // replaced with a set of all the prefixes of specifier resolved. For each
+    // such prefix that exists in the new import map's imports section, we remove
+    // it from that section.
+    for (auto specifier : m_toplevelResolvedModuleSet) {
```
Ditto.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+        message.append("An import map rule for specifier '"_s);
+        message.append(specifier);
+        message.append("' was removed, as it conflicted with already resolved module specifiers."_s);
+        reporter.reportWarning(message.toString());
```
Ditto.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+    // is code unit less than a’s key.</spec>
+    ASSERT(m_scopesVector.isEmpty());
+    for (auto key : m_scopesMap.keys())
+        m_scopesVector.append(key);
```
Ditto.

yoavweiss replied:
Made `auto` a reference, if that's what you meant.

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+
+static Vector<AtomString> findURLPrefixes(String specifier)
+{
+    Vector<size_t> positions;
```
Maybe we should have some inline capacity.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +288,200 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+        position = specifier.find(slash, position);
+    }
+
+    Vector<AtomString> result;
```
Ditto.

yoavweiss replied:
Made `pos` below a reference, if that's what you meant

Constellation reviewed (changes requested):
Algorithm looks fine. But overall, there are too many inefficiencies. Please revise overall code to avoid unnecessary copies / repeated AtomString conversion etc.

yoavweiss reviewed (commented):
Thanks!!

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.h:
```diff
@@ -30,47 +30,63 @@
+    // a new map, by performing more work at addModuleToResolvedModuleSet time,
+    // and by keeping more prefixes in memory.
+    HashSet<AtomString> m_toplevelResolvedModuleSet;
+    HashMap<AtomString, HashSet<AtomString>> m_scopedResolvedModuleMap;
```
Use `UncheckedHashMap` in JSC.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -34,9 +34,37 @@
+            // 2.1.2. Continue.
+            continue;
+        }
+        auto url = newMap.get(specifier);
```
It is looking up twice (keys and look up). We should just iterate key-value.
```
for (auto& [specifier, url] : newMap) {
   ...
}
```

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -34,9 +34,37 @@
+        }
+        auto url = newMap.get(specifier);
+        // 2.2. Set mergedMap[specifier] to url.
+        oldMap.add(specifier, WTFMove(url));
```
And if we use `for (auto& [specifier, url] : newMap)`, then `url` is reference so we should not use `WTFMove` in that case.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -34,9 +34,37 @@
+ImportMap::ImportMap(SpecifierMap&& imports, ScopesMap&& scopesMap, IntegrityMap&& integrity)
+    : m_imports(imports), m_scopesMap(scopesMap), m_integrity(integrity)
+{
+    initializeScopesVector();
```
Since `initializeScopesVector` is not called from anywhere, we should have that code in this constructor instead of extracting it out as `initializeScopesVector` (since it may feel like it can be called from somewhere else).

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +284,189 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+
+    // 3. For each scopePrefix → scopeImports of newImportMapScopes:
+    for (auto& scope : newImportMapScopes) {
+        ImportMap::SpecifierMap& scopeImports = scope.value;
```
After this loop, we no longer need `scopeImports`. So let's just do,
```
auto scopeImports = WTFMove(scope.value);
```

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +284,189 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+                reporter.reportWarning(makeString("An import map scope rule for specifier '"_s, specifier, "' was removed, as it conflicted with already resolved module specifiers."_s));
+                // 3.1.2.1.2. Remove scopeImports[specifier].
+                scopeImports.remove(specifier);
+            }
```
Use `scopeImports.removeIf`. It can make these handling simpler.
So then we do not need `specifiersToRemove`.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +284,189 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+            continue;
+        }
+        // 4.2 Set oldImportMap's integrity[url] to integrity.
+        m_integrity.set(url, newIntegrityValue);
```
Use `auto iter = m_integrity.add(url, newIntegrityValue)`. It will add that value only when `m_integrity` does not have that. And by checking `iter.isNewEntry`, you can detect whether it was containing the entry before. (So, you can avoid double-lookup here).

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +284,189 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+        // developer console.
+        reporter.reportWarning(makeString("An import map rule for specifier '"_s, specifier, "' was removed, as it conflicted with already resolved module specifiers."_s));
+        // 5.2. Remove newImportMapImports[specifier].
+        newImportMapImports.remove(specifier);
```
Avoid double-lookup. (`newImportMapImports.contains` and `newImportMapImports.remove` are looking up twice). Use `iter` from `find` and use that `iter` for removal.

yoavweiss replied:
done!

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +284,189 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+    // is code unit less than a’s key.</spec>
+    ASSERT(m_scopesVector.isEmpty());
+    for (auto& key : m_scopesMap.keys())
+        m_scopesVector.append(key);
```
If it is empty, why not doing `m_scopesVector = copyToVector(m_scopesMap.keys());`?

yoavweiss replied:
I wasn't aware of it. done!

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +284,189 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+        position = specifier.find(slash, position);
+    }
+
+    Vector<AtomString> result;
```
Let's call `result.reserveInitialCapacity(positions.size())`;

yoavweiss replied:
done!

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -254,18 +284,189 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+    // find all the prefixes that the referring script URL has, and add all the
+    // prefixes to the sets of these referring prefixes in the
+    // scoped_resolved_module_map.
+    AtomString atomSpecifier = AtomString(specifier);
```
`specifier` is already `AtomString`.

yoavweiss replied:
done

Constellation commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -34,9 +34,44 @@
+    // 2. For each specifier → url of newMap:
+    for (auto& [specifier, url] : newMap) {
+        // 2.1. If specifier exists in oldMap, then:
+        if (oldMap.contains(specifier)) {
```
Please avoid double lookup. Use add and not use contains here. (do not look up the table with the same twice. Use add and isNewEntry).

Please check these issues throughout the code.

yoavweiss replied:
thanks for catching that, done!

yoavweiss committed:
Implement multiple import maps

https://bugs.webkit.org/show_bug.cgi?id=279025

Reviewed by Yusuke Suzuki.

This PR enables multiple import maps to co-exist in a single document,
and to be loaded after a module script was loaded.

It also removes some code related to external import maps, which was
removed at https://github.com/WebKit/WebKit/pull/37629.

* LayoutTests/TestExpectations: Remove skip and send console to stderr.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/dynamic-import-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/script-tag-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/script-tag-inline-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare-specifiers.sub-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target-dynamic.sub-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target.sub-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/hash-failure-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/nonce-failure-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_data-url-prefix.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_empty-import-map.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_overlapping-entries.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_packages-via-trailing-slashes.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_resolving-null.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_scopes-exact-vs-prefix.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_tricky-specifiers.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_url-specifiers-schemes.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving_url-specifiers.json-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-url-specifiers.sub-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/http-url-like-specifiers.sub-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/already-resolved-dropped-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/basic-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/conflict-first-persists-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/resolution-consistency-in-module-tree-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/resolution-consistency-in-module-tree-inline-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/resolution-consistency-in-module-tree.html: Avoid reliance on async inline module timing.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/url-resolution-conflict-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/with-errors-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-as-classic-script-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/dynamic-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/external-script-bare-descendent-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/external-script-descendent-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/external-script-descendent-with-scope-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/failed-resolution-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/integrity-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/prefix-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/script-descendent-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-overridden/url-resolution-conflict-expected.txt: progression.
* LayoutTests/imported/w3c/web-platform-tests/resources/out-of-scope-test.js: Added. Missed during import.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/import-maps/bare-specifiers.sub-expected.txt: progression.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target-dynamic.sub-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target.sub-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/import-maps/data-url-specifiers.sub-expected.txt: progression.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/import-maps/http-url-like-specifiers.sub-expected.txt: progression.
* LayoutTests/platform/win/imported/w3c/web-platform-tests/import-maps/bare-specifiers.sub-expected.txt: progression.
* LayoutTests/platform/win/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target-dynamic.sub-expected.txt: Removed.
* LayoutTests/platform/win/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target.sub-expected.txt: Removed.
* LayoutTests/platform/win/imported/w3c/web-platform-tests/import-maps/data-url-specifiers.sub-expected.txt: progression.
* LayoutTests/platform/win/imported/w3c/web-platform-tests/import-maps/http-url-like-specifiers.sub-expected.txt: progression.
* Source/JavaScriptCore/builtins/BuiltinNames.h: remove importMapStatus.
* Source/JavaScriptCore/builtins/ModuleLoader.js: remove
  waiting on importMapStatus.
(visibility.PrivateRecursive.async loadModule):
(visibility.PrivateRecursive.async loadAndEvaluateModule):
(visibility.PrivateRecursive.async requestImportModule):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h: remove importMapStatus.
* Source/JavaScriptCore/runtime/ImportMap.cpp: Added multiple import map
  logic.
(JSC::ImportMapInternal::mergeModuleSpecifierMaps): https://html.spec.whatwg.org/C#merge-module-specifier-maps
(JSC::ImportMap::ImportMap): Constructor to enable creation through
a static function.
(JSC::ImportMap::resolve): refactor and add a call to
addModuleToResolvedModuleSet.
(JSC::normalizeSpecifierKey): Change the reporter to a reference.
(JSC::sortAndNormalizeSpecifierMap): Change the reporter to a reference.
(JSC::ImportMap::parseImportMapString): a static function creating
a new ImportMap from an import map string. Replaces registerImportMap.
(JSC::ImportMap::mergeExistingAndNewImportMaps): https://html.spec.whatwg.org/C#merge-existing-and-new-import-maps
the scopes map. The scopes vector is then used to iterate over the
scopes in order.
(JSC::findUrlPrefixes): Find all the prefixes of a URL
(JSC::ImportMap::addModuleToResolvedModuleSet): https://html.spec.whatwg.org/C#add-module-to-resolved-module-set
(JSC::ImportMap::resolve const): Deleted.
(JSC::ImportMap::registerImportMap): Deleted.
* Source/JavaScriptCore/runtime/ImportMap.h: Replace the previous Scopes
  data structure with a HashMap and a Vector, used for fast retrival and
  in-order iteration.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp: Remove isAcquiring
  and pendingImportMap logic.
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::isAcquiringImportMaps const): Deleted.
(JSC::JSGlobalObject::visitChildrenImpl):
(JSC::JSGlobalObject::setAcquiringImportMaps): Deleted.
(JSC::JSGlobalObject::setPendingImportMaps): Deleted.
(JSC::JSGlobalObject::clearPendingImportMaps): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::importMap):
(JSC::JSGlobalObject::importMapStatusPromise const): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp: Remove
  importMapStatus.
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.h:
* Source/WebCore/bindings/js/ScriptController.cpp: Remove isAcquiring
  and pendingImportMap logic. Add error reporting to
  ImportMapLogReporter. Call mergeExistingAndNewImportMaps.
(WebCore::ScriptController::isAcquiringImportMaps): Deleted.
(WebCore::ScriptController::registerImportMap):
(WebCore::ScriptController::setAcquiringImportMaps): Deleted.
(WebCore::ScriptController::setPendingImportMaps): Deleted.
(WebCore::ScriptController::clearPendingImportMaps): Deleted.
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp: Remove spurious
  include.
* Source/WebCore/dom/LoadableClassicScript.cpp:
* Source/WebCore/dom/ScriptElement.cpp: Remove logic firing an error
  event related to isAcquiring. Remove pendingImportMap logic.
(WebCore::ScriptElement::determineScriptType):
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::registerImportMap):
(WebCore::ScriptElement::executePendingScript):
(WebCore::ScriptElement::requestImportMap): Deleted.
* Source/WebCore/dom/ScriptType.h: Fix up spec comments.

Canonical link: https://commits.webkit.org/288358@main
